validator/src/bootstrap.rs

```diff
@@ -616,629 +500,297 @@ mod without_incremental_snapshots {
-    ) -> Option<(Slot, Hash)> {
-        snapshot_utils::get_highest_full_snapshot_archive_info(full_snapshot_archives_dir)
-            .map(|full_snapshot_info| (full_snapshot_info.slot(), *full_snapshot_info.hash()))
+/// Get peer snapshot hashes
+///
+/// The result is a vector of peers with snapshot hashes that:
+/// 1. match a snapshot hash from the known validators
+/// 2. have the highest incremental snapshot slot
+/// 3. have the highest full snapshot slot of (2)
+fn get_peer_snapshot_hashes(
+    cluster_info: &ClusterInfo,
+    validator_config: &ValidatorConfig,
+    rpc_peers: &[ContactInfo],
+    incremental_snapshot_fetch: bool,
+) -> Vec<PeerSnapshotHash> {
+    let mut peer_snapshot_hashes =
+        get_eligible_peer_snapshot_hashes(cluster_info, rpc_peers, incremental_snapshot_fetch);
+    if validator_config.known_validators.is_some() {
+        let known_snapshot_hashes = get_snapshot_hashes_from_known_validators(
+            cluster_info,
+            validator_config,
+            incremental_snapshot_fetch,
+        );
+        retain_peer_snapshot_hashes_that_match_known_snapshot_hashes(
+            &known_snapshot_hashes,
+            &mut peer_snapshot_hashes,
+        );
     }
-}
+    if incremental_snapshot_fetch {
+        retain_peer_snapshot_hashes_with_highest_incremental_snapshot_slot(
+            &mut peer_snapshot_hashes,
+        );
+    }
+    retain_peer_snapshot_hashes_with_highest_full_snapshot_slot(&mut peer_snapshot_hashes);
 
-mod with_incremental_snapshots {
-    use super::*;
+    peer_snapshot_hashes
+}
 
-    /// Struct to wrap the return value from get_rpc_node().  The `rpc_contact_info` is the peer to
-    /// download from, and `snapshot_hash` is the (optional) full and (optional) incremental
-    /// snapshots to download.
-    #[derive(Debug)]
-    struct GetRpcNodeResult {
-        rpc_contact_info: ContactInfo,
-        snapshot_hash: Option<SnapshotHash>,
-    }
+/// Map full snapshot hashes to a set of incremental snapshot hashes.  Each full snapshot hash
+/// is treated as the base for its set of incremental snapshot hashes.
+type KnownSnapshotHashes = HashMap<(Slot, Hash), HashSet<(Slot, Hash)>>;
+
+/// Get the snapshot hashes from known validators.
+///
+/// The snapshot hashes are put into a map from full snapshot hash to a set of incremental
+/// snapshot hashes.  This map will be used as the "known snapshot hashes"; when peers are
+/// queried for their individual snapshot hashes, their results will be checked against this
+/// map to verify correctness.
+///
+/// NOTE: Only a single snashot hash is allowed per slot.  If somehow two known validators have
+/// a snapshot hash with the same slot and _different_ hashes, the second will be skipped.
+/// This applies to both full and incremental snapshot hashes.
+fn get_snapshot_hashes_from_known_validators(
+    cluster_info: &ClusterInfo,
+    validator_config: &ValidatorConfig,
+    incremental_snapshot_fetch: bool,
```

Comment:
Minor/Preference: Same comment here w.r.t. passing in `bootstrap_config` instead of the `incremental_snapshot_fetch` flag.